<a href="https://colab.research.google.com/github/SriAmbica11/SriAmbica_INFO5731_Spring2025/blob/main/Sangineedi_SriAmbica_Assignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment - 1**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.


1.  Question 1-5 using AI assistance is strictly prohibted
2.  Question 6-7 AI assitance is recommended by following guidelines.

Guidelines and expectations file can be found in week 2 module.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [18]:
print('r\t V')
r = [2,4,6,8,10,12,14,16,18,20]
pi = 3.14
for i in r:
  V=(2/3)*pi*(i*i*i)
  V=round(V,2)
  print(i, '\t', V)

r	 V
2 	 16.75
4 	 133.97
6 	 452.16
8 	 1071.79
10 	 2093.33
12 	 3617.28
14 	 5744.11
16 	 8574.29
18 	 12208.32
20 	 16746.67


# Question 2 (10 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [14]:
import numpy as np

# (a) Generate 12 random numbers between 50 and 150
ran_seq = np.random.randint(50, 150, size=12)
print('(a) Generated sequence of 12 random numbers are:', ran_seq)

# (b) Sort the numbers in ascending order
asc_seq = np.sort(ran_seq)
print("\n")
print('(b) Sorted generated sequence in ascending order:', asc_seq)

# (c) Sort the numbers in descending order
desc_seq = asc_seq[::-1]
print("\n")
print('(c) Sorted generated sequence in descending order:', desc_seq)

# (d) Find the average of all numbers
avg_seq = np.mean(ran_seq)
print("\n")
print('(d) Average of the generated sequence:', round(avg_seq, 2))

# (e) Find the average of the first 6 numbers from the ascending order
avg_f_six_asc = np.mean(asc_seq[:6])
# (e) Find the average of the first 6 numbers from the descending order
avg_f_six_desc = np.mean(desc_seq[:6])
print("\n")
print('(e) Average of the first 6 numbers (ascending order):', round(avg_f_six_asc , 2))
print("\n")
print('(e) Average of the first 6 numbers (descending order):', round(avg_f_six_desc, 2))


(a) Generated sequence of 12 random numbers are: [106  55  88 122  53 115  85 143  73  97  95 112]


(b) Sorted generated sequence in ascending order: [ 53  55  73  85  88  95  97 106 112 115 122 143]


(c) Sorted generated sequence in descending order: [143 122 115 112 106  97  95  88  85  73  55  53]


(d) Average of the generated sequence: 95.33


(e) Average of the first 6 numbers (ascending order): 74.83


(e) Average of the first 6 numbers (descending order): 115.83


# Question 3 (10 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [108]:
import string

def split_with_punctuation(sentence):
    words = []
    word = ""
    for char in sentence:
        if char.isalnum():
            word += char
        else:
            if word:
                words.append(word)
                word = ""
            if char.strip():
                words.append(char)
    if word:
        words.append(word)
    return words

sent1 = input("Enter the first sentence: ")
sent2 = input("Enter the second sentence: ")

sent1_list = split_with_punctuation(sent1)
sent2_list = split_with_punctuation(sent2)
comb_list = sent1_list + sent2_list

print("\nThe Combined list of words are:", comb_list)

com_words = set(sent1_list).intersection(set(sent2_list))
print("\nThe Common words in both sentences are:", list(com_words))

def remove_punctuation(sentence):
    return sentence.translate(str.maketrans('', '', string.punctuation))

clean_sent1 = remove_punctuation(sent1)
clean_sent2 = remove_punctuation(sent2)
print("\nThe Sanitized Sentence 1 is:", clean_sent1)
print("The Sanitized Sentence 2 is:", clean_sent2)

def count_vowels(sentence):
    vowels = 'aeiouAEIOU'
    vwl_cnt = {v: sentence.lower().count(v) for v in vowels}
    return vwl_cnt

vwls_sent1 = count_vowels(clean_sent1)
vwls_sent2 = count_vowels(clean_sent2)
print("\nThe Vowel counts for Sentence 1 are:", vwls_sent1)
print("The Vowel counts for Sentence 2 are:", vwls_sent2)


Enter the first sentence: Hello, what is your name?
Enter the second sentence: Hi, my name is Ambica!

The Combined list of words are: ['Hello', ',', 'what', 'is', 'your', 'name', '?', 'Hi', ',', 'my', 'name', 'is', 'Ambica', '!']

The Common words in both sentences are: [',', 'name', 'is']

The Sanitized Sentence 1 is: Hello what is your name
The Sanitized Sentence 2 is: Hi my name is Ambica

The Vowel counts for Sentence 1 are: {'a': 2, 'e': 2, 'i': 1, 'o': 2, 'u': 1, 'A': 0, 'E': 0, 'I': 0, 'O': 0, 'U': 0}
The Vowel counts for Sentence 2 are: {'a': 3, 'e': 1, 'i': 3, 'o': 0, 'u': 0, 'A': 0, 'E': 0, 'I': 0, 'O': 0, 'U': 0}


# Question 4 (10 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [46]:
import math
def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero is not allowed."
    except (SyntaxError, NameError):
        return "Error: Invalid input expression."
    except Exception as e:
        return f"Error: {e}"
def main():
    print("Welcome to the Simple Calculator")
    print("Supports operations: +, -, *, /, %, **, math.sqrt()")
    print("Type 'exit' to quit the calculator.\n")
    while True:
        expression = input("Enter a mathematical expression: ")
        if expression.lower() == "exit":
            print("Exiting calculator.")
            break
        result = evaluate_expression(expression)
        print(f"Result: {result}\n")
if __name__ == "__main__":
    main()

Welcome to the Simple Calculator
Supports operations: +, -, *, /, %, **, math.sqrt()
Type 'exit' to quit the calculator.

Enter a mathematical expression: 5+1
Result: 6

Enter a mathematical expression: 6-3
Result: 3

Enter a mathematical expression: 4*5
Result: 20

Enter a mathematical expression: 36/3
Result: 12.0

Enter a mathematical expression: 7%5
Result: 2

Enter a mathematical expression: 6**2
Result: 36

Enter a mathematical expression: math.sqrt(9)
Result: 3.0

Enter a mathematical expression: 8/0
Result: Error: Division by zero is not allowed.

Enter a mathematical expression: exit
Exiting calculator.


# Question 5 (20 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (5 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (5 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [47]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [101]:
import numpy as np
import pandas as pd
from scipy import stats

# (a) Determine the change in the number of Occupied Housing Units between consecutive years

df = pd.read_csv('/content/Assignment1_denton_housing.csv')
df1 = df[df['title_field'] == 'Occupied Housing Units'].sort_values(by=['year'])
df1['Difference'] = abs(df1['value'].diff())
df1['Year_Range'] = df1['year'].astype(str) + '-' + df1['year'].shift(-1).fillna(df1['year'].max() + 1).astype(int).astype(str)
df1['Difference'] = df1['Difference'].fillna(0).astype(int)
print("\n (a) Yearly difference in the number of occupied housing units (ascending order by year):\n")
print(df1[['year', 'Difference', 'Year_Range']], '\n\n')

# (b) Calculate the percentage of housing units that are vacant and occupied for each year

df2 = df.pivot_table(index='year', columns='title_field', values='value', aggfunc='sum')
df2['Vacant Housing Units'] = (df2['Vacant Housing Units'] / df2['Total Housing Units']) * 100
df2['Occupied Housing Units'] = (df2['Occupied Housing Units'] / df2['Total Housing Units']) * 100
df2['Vacant Housing Units'] = round(df2['Vacant Housing Units'], 2).astype(str) + '%'
df2['Occupied Housing Units'] = round(df2['Occupied Housing Units'], 2).astype(str) + '%'
df2.reset_index(inplace=True)
print("\n (b) Year-wise distribution of percentage for vacant and occupied housing units:\n")
print(df2[['year', 'Vacant Housing Units', 'Occupied Housing Units']], '\n\n')

# (c) Calculate and print the years with the highest number of vacant and occupied housing units

df3 = df.pivot_table(index='year', columns='title_field', values='value', aggfunc='max')
df3 = df3.drop(['Median Gross Rent','Owner Occupied','Renter Occupied','Total Housing Units'], axis=1)
vacant_max_year = df3['Vacant Housing Units'].idxmax()
vacant_max_value = df3['Vacant Housing Units'].max()
occupied_max_year = df3['Occupied Housing Units'].idxmax()
occupied_max_value = df3['Occupied Housing Units'].max()
print("\n (c) Summary of the years with the highest number of vacant and occupied housing units:\n")
print(f"""                          Year   Value
Vacant Housing Units      {vacant_max_year}   {vacant_max_value}
Occupied Housing Units    {occupied_max_year}   {occupied_max_value}
""")

# (d) Calculate and print the harmonic mean of the total housing units

df4 = df[df['title_field'] == 'Total Housing Units']
print(' (d) The harmonic mean of total housing units is', stats.hmean(df4['value']))



 (a) Yearly difference in the number of occupied housing units (ascending order by year):

    year  Difference Year_Range
20  2008           0  2008-2009
17  2009         795  2009-2010
14  2010        5091  2010-2011
11  2011         303  2011-2012
8   2012        1404  2012-2013
5   2013        1565  2013-2014
2   2014        2622  2014-2015 



 (b) Year-wise distribution of percentage for vacant and occupied housing units:

title_field  year Vacant Housing Units Occupied Housing Units
0            2008               10.88%                 89.12%
1            2009                9.07%                 90.93%
2            2010                8.14%                 91.86%
3            2011                9.38%                 90.62%
4            2012                6.68%                 93.32%
5            2013                 8.8%                  91.2%
6            2014                5.73%                 94.27% 



 (c) Summary of the years with the highest number of vacant and oc

# **For question 6 and 7 you can take AI assitance to solve.**

Make sure you follow the instructions given in class to answer these questions.


**Important points:**

1.   You must come up with at least 10 edge cases.
2.   You should provide prompts and revised prompts given to ChatGPT.
2.   Debug your code and write comments for each block of code
3.   Analyzing error will not consider as edge case.

Edge cases are like real life scienarios which are realated to the given question.

# Question 6 (20 points).
Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

**Prompt 1:** Create a Python program to manage a company's product inventory. It should have features to add, remove, and update product details, as well as search for items. The program should also work with CSV files to handle restocking automatically and generate reports showing product values by category. Make sure it can handle problems like missing or corrupted data in the CSV files. Lastly, keep an inventory of 50 products from different categories in the system.










In [109]:
import csv
import os
import random

# Constants
CSV_FILE = "inventory.csv"
RESTOCK_THRESHOLD = 5  # If stock < 5, restock automatically
RESTOCK_AMOUNT = 20  # Amount to restock
CATEGORIES = ["Electronics", "Clothing", "Groceries", "Home", "Toys"]

# Random Product Names
PRODUCT_NAMES = {
    "Electronics": ["Laptop", "Smartphone", "Tablet", "Headphones", "Monitor"],
    "Clothing": ["T-Shirt", "Jeans", "Jacket", "Sneakers", "Cap"],
    "Groceries": ["Milk", "Bread", "Eggs", "Rice", "Cereal"],
    "Home": ["Sofa", "Table", "Chair", "Lamp", "Curtains"],
    "Toys": ["Action Figure", "Board Game", "Puzzle", "Doll", "RC Car"]
}

# Generate Random Products
def generate_random_products(num_products=50):

    '''
    Creates a Products list and generates the random data for the inventory using random module

    Args: we need to give the number of records to be generated as an argument
    '''
    products = []
    for i in range(num_products):
        category = random.choice(CATEGORIES)
        name = random.choice(PRODUCT_NAMES[category])
        product_id = random.randint(1000, 9999)
        stock = random.randint(1, 30)
        price = round(random.uniform(5, 1000), 2)  # Prices between $5 and $1000
        products.append([product_id, name, category, stock, price])
    return products

# Initialize CSV with random data
def initialize_inventory():

    '''
    Checks if the inventory CSV file exists or not. If the CSV file doesn't exists then it calls the function
    generate_random_products() to generate the random data.

    The random generated data will be writed in to the CSV file. Thus, this function initializes the inventory file.
    '''
    if not os.path.exists(CSV_FILE):
        products = generate_random_products()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
            writer.writerows(products)
        print("Inventory initialized with randomly generated data.")

# Load inventory from CSV
def load_inventory():

    '''
    This loads the CSV data in to a list of dictionaries to perform other operations.

    If any exception occurs while reading the file like file is corrupted or unreadable file
    then it automatically created a new file calling initialize_inventory()
    '''
    try:
        with open(CSV_FILE, "r") as file:
            reader = csv.DictReader(file)
            return list(reader)
    except Exception:
        print("Error reading inventory file. Creating a new one.")
        initialize_inventory()
        return load_inventory()

# Save inventory back to CSV
def save_inventory(inventory):

    '''
    Once all the changes like modifying, adding or deleting any product information,
    these changes will be tracked in list of dictionaries that was created
    So, to reflec those changes in the origial CSV file this fuction overwrites the data in to CSV file.
    '''

    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
        for item in inventory:
            writer.writerow([item["ID"], item["Name"], item["Category"], item["Stock"], item["Price"]])

# Add new product
def add_product():

    '''
    If we want to add any new product to our inventory then we can do by using this function.
    it takes runtme inputs for the product details and adds the product in to inventory.

    Raises:
    ValueError: when invalid input is given by user like entering string literal for numerical values.
    '''

    try:
        inventory = load_inventory()
        new_id = int(input("Enter Product ID: "))
        name = input("Enter Product Name: ")
        category = input(f"Enter Category {CATEGORIES}: ")
        stock = int(input("Enter Stock Quantity: "))
        price = float(input("Enter Price: "))

        if category not in CATEGORIES:
            print("Invalid category. Product not added.")
            return

        inventory.append({"ID": new_id, "Name": name, "Category": category, "Stock": stock, "Price": price})
        save_inventory(inventory)
        print(f"Product '{name}' added successfully!")
    except ValueError:
        print("Invalid input! Please enter correct values.")


# Delete product
def delete_product():

    '''
    This function takes the produc id that has to be deleted durig runtime from user and removes the product from inventory.

    If the product id given id not found then it just prints "Product ID not found"

    '''

    inventory = load_inventory()
    product_id = input("Enter Product ID to delete: ")

    new_inventory = [item for item in inventory if item["ID"] != product_id]

    if len(new_inventory) == len(inventory):
        print("Product ID not found.")
    else:
        save_inventory(new_inventory)
        print("Product deleted successfully.")

# Modify product
def modify_product():

    '''
    This is used to modify any details of the product and save the inventory in to a CSV file
    '''
    inventory = load_inventory()
    product_id = input("Enter Product ID to modify: ")

    for item in inventory:
        if item["ID"] == product_id:
            item["Name"] = input("Enter new Name: ") or item["Name"]
            item["Category"] = input("Enter new Category: ") or item["Category"]
            item["Stock"] = input("Enter new Stock: ") or item["Stock"]
            item["Price"] = input("Enter new Price: ") or item["Price"]
            save_inventory(inventory)
            print("Product modified successfully.")
            return

    print("Product ID not found.")

# Search product
def search_product():

    '''This fundtion is used to search for any products. The search should be performed irrespective of alphabetical case.
    So, the input as well as the names in file are converted in to lower case while comparision.
    '''

    inventory = load_inventory()
    search_term = input("Enter product name or category: ").lower()
    results = [item for item in inventory if search_term in item["Name"].lower() or search_term in item["Category"].lower()]

    if results:
        print("Search Results:")
        for item in results:
            print(item)
    else:
        print("No matching products found.")

# Restock items automatically
def auto_restock():
    '''
    It checks if any of the product in the inventory is below threshold level.
    If yes then that product will be restocked automatically
    '''
    inventory = load_inventory()
    restocked_items = []

    for item in inventory:
        if int(item["Stock"]) < RESTOCK_THRESHOLD:
            item["Stock"] = str(int(item["Stock"]) + RESTOCK_AMOUNT)
            restocked_items.append(item["Name"])

    save_inventory(inventory)

    if restocked_items:
        print("Restocked:", ", ".join(restocked_items))
    else:
        print("No restocking needed.")

# Generate report
def generate_report():

    '''
    This generated the total and categorical asset report for the inventory.
    it products the quantity and price to calculates the total stock value we have.
    '''

    inventory = load_inventory()
    category_values = {}

    total_value = 0
    for item in inventory:
        value = int(item["Stock"]) * float(item["Price"])
        total_value += value
        category_values[item["Category"]] = category_values.get(item["Category"], 0) + value

    print("\n=== Inventory Report ===")
    print(f"Total Inventory Value: ${total_value:.2f}")
    print("Category Breakdown:")
    for category, value in category_values.items():
        print(f"{category}: ${value:.2f}")

# Main menu
def main():
    '''
    In the main function, a sequnce of numbers will  be shown to select ay operation
    based on the number selection, the operation will be done to the inventory
    '''
    initialize_inventory()

    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Delete Product")
        print("3. Modify Product")
        print("4. Search Product")
        print("5. Restock Items")
        print("6. Generate Report")
        print("7. Exit")

        choice = input("Enter your choice: ")

        if choice == "1":
            add_product()
        elif choice == "2":
            delete_product()
        elif choice == "3":
            modify_product()
        elif choice == "4":
            search_product()
        elif choice == "5":
            auto_restock()
        elif choice == "6":
            generate_report()
        elif choice == "7":
            print("Exiting... Goodbye!")
            break
        else:
            print("Invalid choice. Try again.")

# Run the program
if __name__ == "__main__":
    main()

Inventory initialized with randomly generated data.

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 6

=== Inventory Report ===
Total Inventory Value: $323131.57
Category Breakdown:
Groceries: $65645.10
Toys: $80862.46
Electronics: $62425.62
Home: $62449.97
Clothing: $51748.42

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 7
Exiting... Goodbye!


**Prompt 2:**
Make sure the system handles the following:

1) No duplicates: Products should not be added more than once to the CSV.
2) Unique categories: A product can't be added under multiple categories.
3) Numeric product ID: The product ID should only have numbers.
4) Detailed reports: Allow generating rep
orts for individual products if requested.
5) No negative stock: Stock quantity must not be negative.
6) Update/delete feedback: Inform the user if they try to update or delete a product that doesn’t exist.
7) Product ID rules: Product IDs should not start with 0.
8) Non-empty name: Product names cannot be empty.
9) Whole number stock: Stock quantity must be a whole number, not a decimal.
10) Unique product ID: Do not add a product if its ID already exists.

In [114]:
import csv
import os
import random

# Constants
CSV_FILE = "inventory.csv"
RESTOCK_THRESHOLD = 5   # If stock < 5, restock automatically
RESTOCK_AMOUNT = 20     # Amount to restock
CATEGORIES = ["Electronics", "Clothing", "Groceries", "Home", "Toys"]

# Random Product Names for Initialization
PRODUCT_NAMES = {
    "Electronics": ["Laptop", "Smartphone", "Tablet", "Headphones", "Monitor"],
    "Clothing": ["T-Shirt", "Jeans", "Jacket", "Sneakers", "Cap"],
    "Groceries": ["Milk", "Bread", "Eggs", "Rice", "Cereal"],
    "Home": ["Sofa", "Table", "Chair", "Lamp", "Curtains"],
    "Toys": ["Action Figure", "Board Game", "Puzzle", "Doll", "RC Car"]
}

# Generate Random Products for Initialization (50 products)
def generate_random_products(num_products=50):
    products = []
    used_ids = set()
    for i in range(num_products):
        category = random.choice(CATEGORIES)
        name = random.choice(PRODUCT_NAMES[category])
        # Ensure unique product IDs during initialization
        while True:
            product_id = random.randint(1001, 9999)  # Avoid starting with 0
            if product_id not in used_ids:
                used_ids.add(product_id)
                break
        stock = random.randint(1, 30)
        price = round(random.uniform(5, 1000), 2)  # Prices between $5 and $1000
        products.append([product_id, name, category, stock, price])
    return products

# Initialize CSV with random data if it doesn't exist
def initialize_inventory():
    if not os.path.exists(CSV_FILE):
        products = generate_random_products()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
            writer.writerows(products)
        print("Inventory initialized with randomly generated data.")

# Load inventory from CSV
def load_inventory():
    try:
        with open(CSV_FILE, "r") as file:
            reader = csv.DictReader(file)
            return list(reader)
    except Exception:
        print("Error reading inventory file. Creating a new one.")
        initialize_inventory()
        return load_inventory()

# Save inventory back to CSV
def save_inventory(inventory):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
        for item in inventory:
            writer.writerow([item["ID"], item["Name"], item["Category"], item["Stock"], item["Price"]])

# Add new product with validations
def add_product():

    '''
    In the upated code, new edge cases are considered

    The product Id should't start with 0
    The newly added product wil be checked against duplicates
    It is validaing if any same product is being added under different category.
    The stock quantity cannot be negative. So, it is being cheked while new insertion
    isdigit() function is being used to check whether the product ID contains all numerical values
    The prodct name shouldn't be empty
    The new product ID id being checked if its lready present in inventory
    the stock should onnly contain the whole numbers.
    '''
    inventory = load_inventory()

    # Input and validate product ID
    product_id_str = input("Enter Product ID (numeric, not starting with 0): ").strip()
    if not product_id_str.isdigit() or product_id_str[0] == "0":
        print("Invalid Product ID. It must be numeric and cannot start with 0.")
        return
    new_id = int(product_id_str)

    # Check if product ID already exists
    for item in inventory:
        if int(item["ID"]) == new_id:
            print("Product ID already exists. Duplicate IDs are not allowed.")
            return

    # Input product name
    name = input("Enter Product Name: ").strip()
    if not name:
        print("Product Name cannot be empty.")
        return

    # Input and validate category
    category = input(f"Enter Category {CATEGORIES}: ").strip()
    if category not in CATEGORIES:
        print("Invalid category. Product not added.")
        return

    # Ensure same product name is not used under a different category
    for item in inventory:
        if item["Name"].lower() == name.lower():
            if item["Category"] != category:
                print(f"Product '{name}' already exists under category '{item['Category']}'.")
                return
            else:
                print(f"Product '{name}' already exists. Duplicate insertion is not allowed.")
                return

    # Input and validate stock
    stock_str = input("Enter Stock Quantity (whole number, non-negative): ").strip()
    if not stock_str.isdigit():
        print("Invalid stock quantity. Please enter a non-negative integer.")
        return
    stock = int(stock_str)

    # Input and validate price
    try:
        price = float(input("Enter Price: ").strip())
    except ValueError:
        print("Invalid price. Please enter a valid number.")
        return

    # Append new product to inventory and save
    inventory.append({
        "ID": new_id,
        "Name": name,
        "Category": category,
        "Stock": stock,
        "Price": price
    })
    save_inventory(inventory)
    print(f"Product '{name}' added successfully!")

# Delete product with validation for non-existing product
def delete_product():

    '''
    While performoig delete operation,
    the validation is performed whether the productID to be deleted is present in inventory or not
    '''
    inventory = load_inventory()
    product_id = input("Enter Product ID to delete: ").strip()

    # Validate product_id input
    if not product_id.isdigit():
        print("Invalid Product ID format.")
        return

    found = False
    new_inventory = []
    for item in inventory:
        if str(item["ID"]) == product_id:
            found = True
        else:
            new_inventory.append(item)

    if not found:
        print("Product ID not found. No deletion performed.")
    else:
        save_inventory(new_inventory)
        print("Product deleted successfully.")

# Modify product with validation for non-existing product and edge cases
def modify_product():

    '''While performoig updates,
    the validation is performed whether the productID to be deleted is present in inventory or not
    conditiond are added to handle dupliactes and same products under different categories.
    -ve stocking issue is addressed aswell while updating
    '''

    inventory = load_inventory()
    product_id = input("Enter Product ID to modify: ").strip()

    # Validate product_id input
    if not product_id.isdigit():
        print("Invalid Product ID format.")
        return

    found = False
    for item in inventory:
        if str(item["ID"]) == product_id:
            found = True
            new_name = input(f"Enter new Name (or press Enter to keep '{item['Name']}'): ").strip()
            if new_name:
                # Check for duplicate name in a different category
                for other in inventory:
                    if other["Name"].lower() == new_name.lower() and other["ID"] != item["ID"]:
                        if other["Category"] != item["Category"]:
                            print(f"Cannot change name to '{new_name}' as it exists under category '{other['Category']}'.")
                            return
                item["Name"] = new_name

            new_category = input(f"Enter new Category (or press Enter to keep '{item['Category']}'): ").strip()
            if new_category:
                if new_category not in CATEGORIES:
                    print("Invalid category. Modification aborted.")
                    return
                item["Category"] = new_category

            new_stock_str = input(f"Enter new Stock (or press Enter to keep '{item['Stock']}'): ").strip()
            if new_stock_str:
                if not new_stock_str.isdigit():
                    print("Invalid stock quantity. It must be a non-negative integer.")
                    return
                new_stock = int(new_stock_str)
                item["Stock"] = new_stock

            new_price_str = input(f"Enter new Price (or press Enter to keep '{item['Price']}'): ").strip()
            if new_price_str:
                try:
                    item["Price"] = float(new_price_str)
                except ValueError:
                    print("Invalid price. Modification aborted.")
                    return

            save_inventory(inventory)
            print("Product modified successfully.")
            break

    if not found:
        print("Product ID not found. No modifications performed.")

# Search product by name or category
def search_product():
    inventory = load_inventory()
    search_term = input("Enter product name or category to search: ").lower().strip()
    results = [item for item in inventory if search_term in item["Name"].lower() or search_term in item["Category"].lower()]

    if results:
        print("\nSearch Results:")
        for item in results:
            print(item)
    else:
        print("No matching products found.")

# Restock items automatically if stock is below threshold
def auto_restock():
    inventory = load_inventory()
    restocked_items = []

    for item in inventory:
        if int(item["Stock"]) < RESTOCK_THRESHOLD:
            new_stock = int(item["Stock"]) + RESTOCK_AMOUNT
            item["Stock"] = new_stock
            restocked_items.append(item["Name"])

    save_inventory(inventory)

    if restocked_items:
        print("Restocked:", ", ".join(restocked_items))
    else:
        print("No restocking needed.")

# Generate report with an option for a detailed product-level report
def generate_report():
    '''
    Report generation is categorized as two types like summarized and detailed
    based on the user run time selection the report is being generated on categorical level or product level
    '''
    inventory = load_inventory()
    report_type = input("Enter report type (summary / detailed): ").strip().lower()

    if report_type == "detailed":
        print("\n=== Detailed Product Report ===")
        for item in inventory:
            value = int(item["Stock"]) * float(item["Price"])
            print(f"ID: {item['ID']}, Name: {item['Name']}, Category: {item['Category']}, "
                  f"Stock: {item['Stock']}, Price: ${float(item['Price']):.2f}, Value: ${value:.2f}")
    else:
        category_values = {}
        total_value = 0
        for item in inventory:
            value = int(item["Stock"]) * float(item["Price"])
            total_value += value
            category = item["Category"]
            category_values[category] = category_values.get(category, 0) + value

        print("\n=== Inventory Report ===")
        print(f"Total Inventory Value: ${total_value:.2f}")
        print("Category Breakdown:")
        for category, value in category_values.items():
            print(f"{category}: ${value:.2f}")

# Main menu
def main():
    initialize_inventory()

    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Delete Product")
        print("3. Modify Product")
        print("4. Search Product")
        print("5. Restock Items")
        print("6. Generate Report")
        print("7. Exit")

        choice = input("Enter your choice: ").strip()

        if choice == "1":
            add_product()
        elif choice == "2":
            delete_product()
        elif choice == "3":
            modify_product()
        elif choice == "4":
            search_product()
        elif choice == "5":
            auto_restock()
        elif choice == "6":
            generate_report()
        elif choice == "7":
            print("Exiting... Goodbye!")
            break
        else:
            print("Invalid choice. Try again.")

# Run the program
if __name__ == "__main__":
    main()


Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 1
Enter Product ID (numeric, not starting with 0): 2547
Enter Product Name: Idly
Enter Category ['Electronics', 'Clothing', 'Groceries', 'Home', 'Toys']: Groceries
Enter Stock Quantity (whole number, non-negative): -10
Invalid stock quantity. Please enter a non-negative integer.

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 1
Enter Product ID (numeric, not starting with 0): OMNDC
Invalid Product ID. It must be numeric and cannot start with 0.

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 1
Enter Product ID (numeric, not starting with 0): 0584
Invalid Product ID. It must be numeric and cannot start with

**Prompt 3:**
I tested the code with all the edge cases, and everything works fine. Now, I want to add user authentication before allowing any changes to the inventory. How can this be done? Provide the code for it.

In [119]:
import csv
import os
import random

# Constants
CSV_FILE = "inventory.csv"
RESTOCK_THRESHOLD = 5   # If stock < 5, restock automatically
RESTOCK_AMOUNT = 20     # Amount to restock
CATEGORIES = ["Electronics", "Clothing", "Groceries", "Home", "Toys"]

# User authentication data (username and password dictionary)
USERS = {
    "User": "India123",
    "manager": "manager456",
    "staff": "staff789"
}

# Random Product Names for Initialization
PRODUCT_NAMES = {
    "Electronics": ["Laptop", "Smartphone", "Tablet", "Headphones", "Monitor"],
    "Clothing": ["T-Shirt", "Jeans", "Jacket", "Sneakers", "Cap"],
    "Groceries": ["Milk", "Bread", "Eggs", "Rice", "Cereal"],
    "Home": ["Sofa", "Table", "Chair", "Lamp", "Curtains"],
    "Toys": ["Action Figure", "Board Game", "Puzzle", "Doll", "RC Car"]
}

# Function to handle user authentication
def authenticate_user():
    print("Please log in to access the inventory system.")
    username = input("Username: ").strip()
    password = input("Password: ").strip()

    if username in USERS and USERS[username] == password:
        print(f"Welcome {username}!")
        return username
    else:
        print("Invalid username or password.")
        return None

# Generate Random Products for Initialization (50 products)
def generate_random_products(num_products=50):
    products = []
    used_ids = set()
    for i in range(num_products):
        category = random.choice(CATEGORIES)
        name = random.choice(PRODUCT_NAMES[category])
        # Ensure unique product IDs during initialization
        while True:
            product_id = random.randint(1001, 9999)  # Avoid starting with 0
            if product_id not in used_ids:
                used_ids.add(product_id)
                break
        stock = random.randint(1, 30)
        price = round(random.uniform(5, 1000), 2)  # Prices between $5 and $1000
        products.append([product_id, name, category, stock, price])
    return products

# Initialize CSV with random data if it doesn't exist
def initialize_inventory():
    if not os.path.exists(CSV_FILE):
        products = generate_random_products()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
            writer.writerows(products)
        print("Inventory initialized with randomly generated data.")

# Load inventory from CSV
def load_inventory():
    try:
        with open(CSV_FILE, "r") as file:
            reader = csv.DictReader(file)
            return list(reader)
    except Exception:
        print("Error reading inventory file. Creating a new one.")
        initialize_inventory()
        return load_inventory()

# Save inventory back to CSV
def save_inventory(inventory):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
        for item in inventory:
            writer.writerow([item["ID"], item["Name"], item["Category"], item["Stock"], item["Price"]])

# Add new product with validations
def add_product():
    inventory = load_inventory()

    # Input and validate product ID
    product_id_str = input("Enter Product ID (numeric, not starting with 0): ").strip()
    if not product_id_str.isdigit() or product_id_str[0] == "0":
        print("Invalid Product ID. It must be numeric and cannot start with 0.")
        return
    new_id = int(product_id_str)

    # Check if product ID already exists
    for item in inventory:
        if int(item["ID"]) == new_id:
            print("Product ID already exists. Duplicate IDs are not allowed.")
            return

    # Input product name
    name = input("Enter Product Name: ").strip()
    if not name:
        print("Product Name cannot be empty.")
        return

    # Input and validate category
    category = input(f"Enter Category {CATEGORIES}: ").strip()
    if category not in CATEGORIES:
        print("Invalid category. Product not added.")
        return

    # Ensure same product name is not used under a different category
    for item in inventory:
        if item["Name"].lower() == name.lower():
            if item["Category"] != category:
                print(f"Product '{name}' already exists under category '{item['Category']}'.")
                return
            else:
                print(f"Product '{name}' already exists. Duplicate insertion is not allowed.")
                return

    # Input and validate stock
    stock_str = input("Enter Stock Quantity (whole number, non-negative): ").strip()
    if not stock_str.isdigit():
        print("Invalid stock quantity. Please enter a non-negative integer.")
        return
    stock = int(stock_str)

    # Input and validate price
    try:
        price = float(input("Enter Price: ").strip())
    except ValueError:
        print("Invalid price. Please enter a valid number.")
        return

    # Append new product to inventory and save
    inventory.append({
        "ID": new_id,
        "Name": name,
        "Category": category,
        "Stock": stock,
        "Price": price
    })
    save_inventory(inventory)
    print(f"Product '{name}' added successfully!")

# Modify an existing product's details
def modify_product():
    inventory = load_inventory()

    product_id_str = input("Enter the Product ID to modify: ").strip()
    if not product_id_str.isdigit():
        print("Invalid Product ID format.")
        return

    product_id = int(product_id_str)
    product_found = False

    for item in inventory:
        if int(item["ID"]) == product_id:
            product_found = True
            print(f"\nCurrent Details of Product ID {product_id}:")
            print(f"Name: {item['Name']}, Category: {item['Category']}, Stock: {item['Stock']}, Price: ${item['Price']:.2f}")

            # Modify name
            new_name = input(f"Enter new name for the product (or press Enter to keep '{item['Name']}'): ").strip()
            if new_name:
                item['Name'] = new_name

            # Modify category
            new_category = input(f"Enter new category for the product (or press Enter to keep '{item['Category']}'): ").strip()
            if new_category and new_category in CATEGORIES:
                item['Category'] = new_category

            # Modify stock
            new_stock_str = input(f"Enter new stock quantity for the product (or press Enter to keep {item['Stock']}): ").strip()
            if new_stock_str:
                if new_stock_str.isdigit():
                    item['Stock'] = int(new_stock_str)
                else:
                    print("Invalid stock quantity.")

            # Modify price
            new_price_str = input(f"Enter new price for the product (or press Enter to keep ${item['Price']}): ").strip()
            if new_price_str:
                try:
                    item['Price'] = float(new_price_str)
                except ValueError:
                    print("Invalid price entered.")

            save_inventory(inventory)
            print(f"Product ID {product_id} updated successfully!")
            break

    if not product_found:
        print("Product ID not found. No modification made.")

# Delete product with validation for non-existing product
def delete_product():
    inventory = load_inventory()
    product_id = input("Enter Product ID to delete: ").strip()

    # Validate product_id input
    if not product_id.isdigit():
        print("Invalid Product ID format.")
        return

    found = False
    new_inventory = []
    for item in inventory:
        if str(item["ID"]) == product_id:
            found = True
        else:
            new_inventory.append(item)

    if not found:
        print("Product ID not found. No deletion performed.")
    else:
        save_inventory(new_inventory)
        print("Product deleted successfully.")

# Search product by name or category
def search_product():
    inventory = load_inventory()
    search_term = input("Enter product name or category to search: ").lower().strip()
    results = [item for item in inventory if search_term in item["Name"].lower() or search_term in item["Category"].lower()]

    if results:
        print("\nSearch Results:")
        for item in results:
            print(item)
    else:
        print("No matching products found.")

# Restock items automatically if stock is below threshold
def auto_restock():
    inventory = load_inventory()
    restocked_items = []

    for item in inventory:
        if int(item["Stock"]) < RESTOCK_THRESHOLD:
            new_stock = int(item["Stock"]) + RESTOCK_AMOUNT
            item["Stock"] = new_stock
            restocked_items.append(item["Name"])

    save_inventory(inventory)

    if restocked_items:
        print("Restocked:", ", ".join(restocked_items))
    else:
        print("No restocking needed.")

# Generate report with an option for a detailed product-level report
def generate_report():
    inventory = load_inventory()
    report_type = input("Enter report type (summary / detailed): ").strip().lower()

    if report_type == "detailed":
        print("\n=== Detailed Product Report ===")
        for item in inventory:
            value = int(item["Stock"]) * float(item["Price"])
            print(f"ID: {item['ID']}, Name: {item['Name']}, Category: {item['Category']}, "
                  f"Stock: {item['Stock']}, Price: ${float(item['Price']):.2f}, Value: ${value:.2f}")
    else:
        category_values = {}
        total_value = 0
        for item in inventory:
            value = int(item["Stock"]) * float(item["Price"])
            total_value += value
            category = item["Category"]
            category_values[category] = category_values.get(category, 0) + value

        print("\n=== Inventory Report ===")
        print(f"Total Inventory Value: ${total_value:.2f}")
        print("Category Breakdown:")
        for category, value in category_values.items():
            print(f"{category}: ${value:.2f}")

# Main menu
def main():
    initialize_inventory()
    # Authenticate user
    username = authenticate_user()
    if username is None:
        print("Access denied. Exiting program.")
        return

    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Delete Product")
        print("3. Modify Product")
        print("4. Search Product")
        print("5. Restock Items")
        print("6. Generate Report")
        print("7. Exit")

        choice = input("Enter your choice: ").strip()

        if choice == "1":
            add_product()
        elif choice == "2":
            delete_product()
        elif choice == "3":
            modify_product()
        elif choice == "4":
            search_product()
        elif choice == "5":
            auto_restock()
        elif choice == "6":
            generate_report()
        elif choice == "7":
            print("Exiting... Goodbye!")
            break
        else:
            print("Invalid choice. Try again.")

# Run the program
if __name__ == "__main__":
    main()

Please log in to access the inventory system.
Username: User
Password: India123
Welcome User!

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 1
Enter Product ID (numeric, not starting with 0): 4589
Enter Product Name: Milk
Enter Category ['Electronics', 'Clothing', 'Groceries', 'Home', 'Toys']: Groceries
Product 'Milk' already exists. Duplicate insertion is not allowed.

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 6
Enter report type (summary / detailed): summary

=== Inventory Report ===
Total Inventory Value: $323131.57
Category Breakdown:
Groceries: $65645.10
Toys: $80862.46
Electronics: $62425.62
Home: $62449.97
Clothing: $51748.42

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate 

# Question 7 (20 points)

Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

**Prompt:1**
Create a Python program for a flight booking system with features to manage around 50 flights. It should allow adding new flights, searching for available trips, booking and canceling seats, and generating performance reports.

In [131]:
import random
import csv
import os

# Constants
CSV_FILE = "flights.csv"
SEAT_CAPACITY = 100  # Each flight has 100 seats
FLIGHT_COUNT = 50

# Sample Cities for Flights
CITIES = ["New York", "Los Angeles", "Chicago", "Dallas", "San Francisco",
          "Miami", "Seattle", "Denver", "Boston", "Houston"]

# Generate random flights
def generate_flights(num_flights=FLIGHT_COUNT):
    '''
    This function is used to generate the list of flights that are available.
    The data is being generated using random module
    It takes number as an argument that tells the number of flights reuired in initial dataset
    '''
    flights = []
    for _ in range(num_flights):
        flight_id = random.randint(1000, 9999)  # Unique Flight ID
        departure = random.choice(CITIES)
        destination = random.choice([city for city in CITIES if city != departure])  # Different city
        price = round(random.uniform(100, 1000), 2)  # Ticket price range
        available_seats = SEAT_CAPACITY  # Initial seat availability
        flights.append([flight_id, departure, destination, price, available_seats])
    return flights

# Initialize flight data
def initialize_flight_system():
    if not os.path.exists(CSV_FILE):
        flights = generate_flights()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["Flight ID", "Departure", "Destination", "Price", "Available Seats"])
            writer.writerows(flights)
        print("Flight system initialized with 50 random flights.")

# Load flights from CSV
def load_flights():
    try:
        with open(CSV_FILE, "r") as file:
            reader = csv.DictReader(file)
            return list(reader)
    except Exception:
        print("Error reading flight file. Initializing new flights.")
        initialize_flight_system()
        return load_flights()

# Save flights back to CSV
def save_flights(flights):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["Flight ID", "Departure", "Destination", "Price", "Available Seats"])
        for flight in flights:
            writer.writerow([flight["Flight ID"], flight["Departure"], flight["Destination"],
                             flight["Price"], flight["Available Seats"]])

# Add a new flight
def add_flight():
    '''
    This function asks for the new flight details to be added in to the list
    and adds the flight details in to the CSV file

    '''
    flights = load_flights()

    flight_id = random.randint(1000, 9999)
    departure = input(f"Enter Departure City {CITIES}: ").strip()
    if departure not in CITIES:
        print("Invalid departure city.")
        return

    destination = input(f"Enter Destination City {CITIES} (must be different from departure): ").strip()
    if destination not in CITIES or destination == departure:
        print("Invalid destination city.")
        return

    try:
        price = float(input("Enter Ticket Price: "))
    except ValueError:
        print("Invalid price format.")
        return

    flights.append({
        "Flight ID": flight_id,
        "Departure": departure,
        "Destination": destination,
        "Price": price,
        "Available Seats": SEAT_CAPACITY
    })

    save_flights(flights)
    print(f"Flight {flight_id} added successfully.")


def find_flights():
    '''
    This function is used to search the flights based on the departure and destination cities.
    '''
    flights = load_flights()
    departure = input("Enter Departure City: ").strip()
    destination = input("Enter Destination City: ").strip()

    results = [flight for flight in flights if flight["Departure"] == departure and flight["Destination"] == destination]

    if results:
        print("\nAvailable Flights:")
        for flight in results:
            print(f"Flight {flight['Flight ID']} - {flight['Departure']} to {flight['Destination']}, Price: ${flight['Price']}, Seats Available: {flight['Available Seats']}")
    else:
        print("No flights found for this route.")

# Book a flight seat
def book_seat():
    '''
    This flighst takes the fight id as the input and if the flight exists in the csv file
    then flight will be booked and the available seats will be reduced by 1
    if the flight is full user will be informed about the unavailability
    '''
    flights = load_flights()
    flight_id = input("Enter Flight ID to book a seat: ").strip()

    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Available Seats"]) > 0:
                flight["Available Seats"] = str(int(flight["Available Seats"]) - 1)
                save_flights(flights)
                print("Seat booked successfully!")
                return
            else:
                print("No seats available for this flight.")
                return

    print("Flight ID not found.")

# Cancel a booking
def cancel_booking():
    '''
    if user requests for a cancellation by entering the flight id
    then the avialable seats will be increased by 1 and booking will be cancelled
    '''
    flights = load_flights()
    flight_id = input("Enter Flight ID to cancel booking: ").strip()

    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Available Seats"]) < SEAT_CAPACITY:
                flight["Available Seats"] = str(int(flight["Available Seats"]) + 1)
                save_flights(flights)
                print("Booking canceled successfully.")
                return
            else:
                print("All seats are already available, no bookings to cancel.")
                return

    print("Flight ID not found.")

# Generate report of flights and bookings
def generate_report():
    '''
    This function is used to generate the statastics report of the flight system
    like total available flights, total revenue generated, bookings etc.
    '''
    flights = load_flights()

    total_flights = len(flights)
    total_seats = total_flights * SEAT_CAPACITY
    booked_seats = sum(SEAT_CAPACITY - int(flight["Available Seats"]) for flight in flights)
    revenue = sum((SEAT_CAPACITY - int(flight["Available Seats"])) * float(flight["Price"]) for flight in flights)

    print("\n=== Flight Booking System Report ===")
    print(f"Total Flights: {total_flights}")
    print(f"Total Seats Available: {total_seats}")
    print(f"Seats Booked: {booked_seats}")
    print(f"Total Revenue: ${revenue:.2f}")

# Main menu function
def main():
    '''
    This is the main function that gives list of numeric values and  asks for the operation to be performed
    Based on the selection the operation will be performed.
    '''
    initialize_flight_system()

    while True:
        print("\nFlight Booking System")
        print("1. Add Flight")
        print("2. Find Flights")
        print("3. Book Seat")
        print("4. Cancel Booking")
        print("5. Generate Report")
        print("6. Exit")

        choice = input("Enter your choice: ").strip()

        if choice == "1":
            add_flight()
        elif choice == "2":
            find_flights()
        elif choice == "3":
            book_seat()
        elif choice == "4":
            cancel_booking()
        elif choice == "5":
            generate_report()
        elif choice == "6":
            print("Exiting... Thank you for using the Flight Booking System!")
            break
        else:
            print("Invalid choice. Try again.")

# Run the program
if __name__ == "__main__":
    main()


Flight Booking System
1. Add Flight
2. Find Flights
3. Book Seat
4. Cancel Booking
5. Generate Report
6. Exit
Enter your choice: 2
Enter Departure City: Dallas
Enter Destination City: New York

Available Flights:
Flight 2001 - Dallas to New York, Price: $393.53, Seats Available: 100

Flight Booking System
1. Add Flight
2. Find Flights
3. Book Seat
4. Cancel Booking
5. Generate Report
6. Exit
Enter your choice: 6
Exiting... Thank you for using the Flight Booking System!


**Prompt-2:**

Consider the following edge cases for the flight booking system:

If the user enters an invalid flight ID, display a clear error message and prompt for a valid ID.
Prevent overbooking when no seats are available.
Ticket prices cannot be negative.
Flight IDs must be unique, with no duplicates.
City names should not be case-sensitive when booking, canceling, or searching for flights.
Flight IDs can have alphanumeric characters and must be exactly 8 characters long.
The flight ID should be entered by the user when adding a flight.
Bookings must include a date and time to avoid confusion.
Users cannot book flights for past dates.








In [139]:
import random
import csv
import os
import datetime
import string

# Constants
CSV_FILE = "flights.csv"
SEAT_CAPACITY = 100  # Each flight has 100 seats
FLIGHT_COUNT = 50

# Sample Cities for Flights
CITIES = ["New York", "Los Angeles", "Chicago", "Dallas", "San Francisco",
          "Miami", "Seattle", "Denver", "Boston", "Houston"]


def generate_flight_id():
    #generate alphanumeric flight id's of length 8
    return ''.join(random.choices(string.ascii_uppercase + string.digits, k=8))

# Validate flight ID format
def is_valid_flight_id(flight_id):
    #alpha numeric and length check for the flight id's
    return len(flight_id) == 8 and flight_id.isalnum()

# Validate date format and ensure it is not in the past
def is_valid_future_date(date_str):
    '''we are making sure that the flight date is the valid future date
    ensures the flight timing is not completed

    Raises value error if invalid dates are entered
    '''
    try:
        flight_date = datetime.datetime.strptime(date_str, "%Y-%m-%d %H:%M")
        return flight_date > datetime.datetime.now()
    except ValueError:
        return False

# Generate flights

def generate_flights(num_flights=FLIGHT_COUNT):
    '''
    The flight data is getting generated using random module
    additionally, date and time are mentioned now for the flights
    Alphanumeric 8 character flight id's are being generated
    '''
    flights = []
    for _ in range(num_flights):
        flight_id = generate_flight_id()
        departure = random.choice(CITIES)
        destination = random.choice([city for city in CITIES if city != departure])
        price = round(random.uniform(100, 1000), 2)
        available_seats = SEAT_CAPACITY
        date = (datetime.datetime.now() + datetime.timedelta(days=random.randint(1, 30))).strftime("%Y-%m-%d %H:%M")
        flights.append([flight_id, departure, destination, price, available_seats, date])
    return flights

# Initialize flight system
def initialize_flight_system():
    if not os.path.exists(CSV_FILE):
        flights = generate_flights()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["Flight ID", "Departure", "Destination", "Price", "Available Seats", "Date"])
            writer.writerows(flights)
        print("Flight system initialized with 50 random flights.")

# Load flights from CSV
def load_flights():
    if not os.path.exists(CSV_FILE):
        initialize_flight_system()
    with open(CSV_FILE, "r") as file:
        reader = csv.DictReader(file)
        return list(reader)

# Save flights to CSV
def save_flights(flights):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["Flight ID", "Departure", "Destination", "Price", "Available Seats", "Date"])
        for flight in flights:
          writer.writerow([flight["Flight ID"], flight["Departure"], flight["Destination"], flight["Price"], flight["Available Seats"], flight["Date"]])

# Add a new flight
def add_flight():
    '''
    Here the user enetered flight is being checked whether it is alphanumeric
    the flight id should be of length 8
    the flight id should be same as the existing flight id
    the flight id being taken from user
    '''
    flights = load_flights()

    flight_id = input("Enter Flight ID (8 alphanumeric characters): ").strip().upper()
    if not is_valid_flight_id(flight_id) or any(f["Flight ID"] == flight_id for f in flights):
        print("Invalid or duplicate Flight ID.")
        return

    departure = input(f"Enter Departure City {CITIES}: ").strip()
    if departure not in CITIES:
        print("Invalid departure city.")
        return

    destination = input(f"Enter Destination City {CITIES} (must be different from departure): ").strip()
    if destination not in CITIES or destination == departure:
        print("Invalid destination city.")
        return

    date = input("Enter flight date and time (YYYY-MM-DD HH:MM): ").strip()
    if not is_valid_future_date(date):
        print("Invalid or past date.")
        return

    '''
    Here we are handling the negative prices

    Raises value error if negatve price ie entered
    '''

    try:
        price = float(input("Enter Ticket Price: "))
        if price < 0:
            raise ValueError("Price cannot be negative.")
    except ValueError:
        print("Invalid price format.")
        return

    flights.append({
        "Flight ID": flight_id,
        "Departure": departure,
        "Destination": destination,
        "Price": price,
        "Available Seats": SEAT_CAPACITY,
        "Date": date
    })
    save_flights(flights)
    print(f"Flight {flight_id} added successfully.")

# Find flights
def find_flights():
    flights = load_flights()
    departure = input("Enter Departure City: ").strip()
    destination = input("Enter Destination City: ").strip()

    results = [flight for flight in flights if flight["Departure"] == departure and flight["Destination"] == destination]
    if results:
        for flight in results:
            # Check if 'Date' key exists before accessing it
            date = flight.get("Date", "N/A")  # Default to "N/A" if 'Date' is missing
            print(f"Flight {flight['Flight ID']} - {flight['Departure']} to {flight['Destination']}, Price: ${flight['Price']}, Seats Available: {flight['Available Seats']}, Date: {date}")
    else:
        print("No flights found for this route.")

# Book seat
def book_seat():
    flights = load_flights()
    flight_id = input("Enter Flight ID to book a seat: ").strip().upper()
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Available Seats"]) > 0:
                flight["Available Seats"] = str(int(flight["Available Seats"]) - 1)
                save_flights(flights)
                print("Seat booked successfully!")
                return
            else:
                print("No seats available for this flight.")
                return
    print("Flight ID not found.")

# Cancel booking
def cancel_booking():
    flights = load_flights()
    flight_id = input("Enter Flight ID to cancel booking: ").strip().upper()
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Available Seats"]) < SEAT_CAPACITY:
                flight["Available Seats"] = str(int(flight["Available Seats"]) + 1)
                save_flights(flights)
                print("Booking canceled successfully.")
                return
            else:
                print("All seats are already available.")
                return
    print("Flight ID not found.")

# Generate report
def generate_report():
    flights = load_flights()
    booked_seats = sum(SEAT_CAPACITY - int(f["Available Seats"]) for f in flights)
    revenue = sum((SEAT_CAPACITY - int(f["Available Seats"])) * float(f["Price"]) for f in flights)
    print(f"Total Flights: {len(flights)}, Seats Booked: {booked_seats}, Total Revenue: ${revenue:.2f}")

# Main menu
def main():
    initialize_flight_system()
    while True:
        choice = input("1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: ").strip()
        if choice == "1": add_flight()
        elif choice == "2": find_flights()
        elif choice == "3": book_seat()
        elif choice == "4": cancel_booking()
        elif choice == "5": generate_report()
        elif choice == "6": break
        else: print("Invalid choice. Try again.")

if __name__ == "__main__":
    main()

1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: 2
Enter Departure City: Houston
Enter Destination City: Chicago
Flight 3678 - Houston to Chicago, Price: $153.41, Seats Available: 100, Date: N/A
1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: 1
Enter Flight ID (8 alphanumeric characters): adsj67889
Invalid or duplicate Flight ID.
1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: 1
Enter Flight ID (8 alphanumeric characters): asdf3456
Enter Departure City ['New York', 'Los Angeles', 'Chicago', 'Dallas', 'San Francisco', 'Miami', 'Seattle', 'Denver', 'Boston', 'Houston']: dallas
Invalid departure city.
1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: 6


The code works well for most cases, but there's a problem with city names. It doesn’t recognize the names correctly if users enter them with different capital or small letters. For example, searching for "new york" won’t match "New York." This affects booking, canceling, and searching for flights.

**Prompt 3:** Fix the code so that it handles city names without worrying about capital or small letters. The program should work no matter how the user types the city name.

In [141]:
import random
import csv
import os
import datetime
import string

# Constants
CSV_FILE = "flights.csv"
SEAT_CAPACITY = 100  # Each flight has 100 seats
FLIGHT_COUNT = 50

# Sample Cities for Flights
CITIES = ["New York", "Los Angeles", "Chicago", "Dallas", "San Francisco",
          "Miami", "Seattle", "Denver", "Boston", "Houston"]


def generate_flight_id():

    #generate alphanumeric flight id's of length 8
    return ''.join(random.choices(string.ascii_uppercase + string.digits, k=8))

# Validate flight ID format
def is_valid_flight_id(flight_id):
    #alpha numeric and length check for the flight id's
    return len(flight_id) == 8 and flight_id.isalnum()

# Validate date format and ensure it is not in the past
def is_valid_future_date(date_str):
    '''we are making sure that the flight date is the valid future date
    ensures the flight timing is not completed

    Raises value error if invalid dates are entered
    '''
    try:
        flight_date = datetime.datetime.strptime(date_str, "%Y-%m-%d %H:%M")
        return flight_date > datetime.datetime.now()
    except ValueError:
        return False

# Generate flights

def generate_flights(num_flights=FLIGHT_COUNT):
    '''
    The flight data is getting generated using random module
    additionally, date and time are mentioned now for the flights
    Alphanumeric 8 character flight id's are being generated
    '''
    flights = []
    for _ in range(num_flights):
        flight_id = generate_flight_id()
        departure = random.choice(CITIES)
        destination = random.choice([city for city in CITIES if city != departure])
        price = round(random.uniform(100, 1000), 2)
        available_seats = SEAT_CAPACITY
        date = (datetime.datetime.now() + datetime.timedelta(days=random.randint(1, 30))).strftime("%Y-%m-%d %H:%M")
        flights.append([flight_id, departure, destination, price, available_seats, date])
    return flights

# Initialize flight system
def initialize_flight_system():
    if not os.path.exists(CSV_FILE):
        flights = generate_flights()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["Flight ID", "Departure", "Destination", "Price", "Available Seats", "Date"])
            writer.writerows(flights)
        print("Flight system initialized with 50 random flights.")

# Load flights from CSV
def load_flights():
    if not os.path.exists(CSV_FILE):
        initialize_flight_system()
    with open(CSV_FILE, "r") as file:
        reader = csv.DictReader(file)
        return list(reader)

# Save flights to CSV
def save_flights(flights):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["Flight ID", "Departure", "Destination", "Price", "Available Seats", "Date"])
        for flight in flights:
          writer.writerow([flight["Flight ID"], flight["Departure"], flight["Destination"], flight["Price"], flight["Available Seats"], flight["Date"]])

# Add a new flight
def add_flight():
    '''
    Here the user enetered flight is being checked whether it is alphanumeric
    the flight id should be of length 8
    the flight id should be same as the existing flight id
    the flight id being taken from user
    the city names case sensiivity also handled here
    '''
    flights = load_flights()

    flight_id = input("Enter Flight ID (8 alphanumeric characters): ").strip().upper()
    if not is_valid_flight_id(flight_id) or any(f["Flight ID"] == flight_id for f in flights):
        print("Invalid or duplicate Flight ID.")
        return

    departure = input(f"Enter Departure City {CITIES}: ").strip().lower()  # Convert to lowercase
    if departure not in [city.lower() for city in CITIES]:  # Compare with lowercase cities list
        print("Invalid departure city.")
        return

    destination = input(f"Enter Destination City {CITIES} (must be different from departure): ").strip().lower()
    if destination not in [city.lower() for city in CITIES] or destination == departure:
        print("Invalid destination city.")
        return

    date = input("Enter flight date and time (YYYY-MM-DD HH:MM): ").strip()
    if not is_valid_future_date(date):
        print("Invalid or past date.")
        return

    '''
    Here we are handling the negative prices

    Raises value error if negatve price ie entered
    '''
    try:
        price = float(input("Enter Ticket Price: "))
        if price < 0:
            raise ValueError("Price cannot be negative.")
    except ValueError:
        print("Invalid price format.")
        return

    flights.append({
        "Flight ID": flight_id,
        "Departure": departure.capitalize(),  # Store cities with proper capitalization
        "Destination": destination.capitalize(),
        "Price": price,
        "Available Seats": SEAT_CAPACITY,
        "Date": date
    })
    save_flights(flights)
    print(f"Flight {flight_id} added successfully.")


# Find flights
def find_flights():
    '''
    This fucntion is used to search the flights
    based on the departure and destination city names
    '''

    flights = load_flights()
    departure = input("Enter Departure City: ").strip().lower()  # Convert to lowercase
    destination = input("Enter Destination City: ").strip().lower()  # Convert to lowercase

    results = [flight for flight in flights if flight["Departure"].lower() == departure and flight["Destination"].lower() == destination]
    if results:
        for flight in results:
            print(f"Flight {flight['Flight ID']} - {flight['Departure']} to {flight['Destination']}, Price: ${flight['Price']}, Seats Available: {flight['Available Seats']}, Date: {flight['Date']}")
    else:
        print("No flights found for this route.")


# Book seat
def book_seat():
    '''
    This flighst takes the fight id as the input and if the flight exists in the csv file
    then flight will be booked and the available seats will be reduced by 1
    if the flight is full user will be informed about the unavailability
    '''
    flights = load_flights()
    flight_id = input("Enter Flight ID to book a seat: ").strip().upper()  # Flight ID is case sensitive, keep it as uppercase
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Available Seats"]) > 0:
                flight["Available Seats"] = str(int(flight["Available Seats"]) - 1)
                save_flights(flights)
                print("Seat booked successfully!")
                return
            else:
                print("No seats available for this flight.")
                return
    print("Flight ID not found.")


# Cancel booking
def cancel_booking():
    '''
    if user requests for a cancellation by entering the flight id
    then the avialable seats will be increased by 1 and booking will be cancelled
    '''
    flights = load_flights()
    flight_id = input("Enter Flight ID to cancel booking: ").strip().upper()
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Available Seats"]) < SEAT_CAPACITY:
                flight["Available Seats"] = str(int(flight["Available Seats"]) + 1)
                save_flights(flights)
                print("Booking canceled successfully.")
                return
            else:
                print("All seats are already available.")
                return
    print("Flight ID not found.")


# Generate report
def generate_report():
    '''
    This function is used to generate the statastics report of the flight system
    like total available flights, total revenue generated, bookings etc.
    '''
    flights = load_flights()
    booked_seats = sum(SEAT_CAPACITY - int(f["Available Seats"]) for f in flights)
    revenue = sum((SEAT_CAPACITY - int(f["Available Seats"])) * float(f["Price"]) for f in flights)
    print(f"Total Flights: {len(flights)}, Seats Booked: {booked_seats}, Total Revenue: ${revenue:.2f}")

# Main menu
def main():
    initialize_flight_system()
    while True:
        choice = input("1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: ").strip()
        if choice == "1": add_flight()
        elif choice == "2": find_flights()
        elif choice == "3": book_seat()
        elif choice == "4": cancel_booking()
        elif choice == "5": generate_report()
        elif choice == "6": break
        else: print("Invalid choice. Try again.")

if __name__ == "__main__":
    main()

1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: 1
Enter Flight ID (8 alphanumeric characters): asfg9862
Enter Departure City ['New York', 'Los Angeles', 'Chicago', 'Dallas', 'San Francisco', 'Miami', 'Seattle', 'Denver', 'Boston', 'Houston']: new york
Enter Destination City ['New York', 'Los Angeles', 'Chicago', 'Dallas', 'San Francisco', 'Miami', 'Seattle', 'Denver', 'Boston', 'Houston'] (must be different from departure): dallas
Enter flight date and time (YYYY-MM-DD HH:MM): 2025-02-26 08:30
Enter Ticket Price: 768
Flight ASFG9862 added successfully.
1. Add Flight 2. Find Flights 3. Book Seat 4. Cancel Booking 5. Generate Report 6. Exit: 6


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [145]:
''' This assignment was pretty tough, but I have learned a lot while doing the assignment. The first few questions which are from 1 to 5 were a mix of easy and tricky ones and I had fun solving them. But questions 6 and 7 gave me a hard time. Even with the help of AI, I still found them confusing and very challenging. Overall, it was difficult but I am glad that I got to learn alot throughout the process. '''

' This assignment was pretty tough, but I have learned a lot while doing the assignment. The first few questions which are from 1 to 5 were a mix of easy and tricky ones and I had fun solving them. But questions 6 and 7 gave me a hard time. Even with the help of AI, I still found them confusing and very challenging. Overall, it was difficult but I am glad that I got to learn alot throughout the process. '